scim2-sdk-common/src/main/java/com/unboundid/scim2/common/utils/JsonUtils.java

```diff
@@ -254,55 +257,131 @@ else if(node.isArray())
+     */
+    private boolean processArrayReplace(@NotNull final ObjectNode parent,
+                                        @NotNull final String field,
+                                        @NotNull final ArrayNode array,
+                                        @NotNull final Filter valueFilter)
+        throws ScimException
+    {
+      boolean nodeUpdated = false;
+
+      // If the 'value' is an empty array, then the client is actually
+      // requesting that we delete values that match the filter.
+      boolean isDelete = isNullNodeOrEmptyArray(value);
+      List<Integer> indexesToDelete = new ArrayList<>();
+
+      for (int i = 0; i < array.size(); i++)
+      {
+        JsonNode attributeValue = array.get(i);
+
+        // We should only perform processing on this inner node if it matches
+        // the filter.
+        if (!FilterEvaluator.evaluate(valueFilter, attributeValue))
         {
-          // in replace mode, a value filter requires that the target node
-          // be an array and that we can find matching value(s)
-          boolean matchesFound = false;
-          if (node.isArray())
-          {
-            for(int i = 0; i < node.size(); i++)
-            {
-              if(FilterEvaluator.evaluate(valueFilter, node.get(i)))
-              {
-                matchesFound = true;
-                if(node.get(i).isObject() && value.isObject())
-                {
-                  updateNode((ObjectNode) node.get(i), null, value);
-                }
-                else
-                {
-                  ((ArrayNode) node).set(i, value);
-                }
-              }
-            }
-          }
-          // exception: this allows filters on singular values if
-          // and only if the filter uses the "value" attribute to
-          // reference the value of the value node.
-          else if (FilterEvaluator.evaluate(valueFilter, node))
-          {
-            matchesFound = true;
-            updateNode(parent, field, value);
-          }
-          if(!matchesFound)
-          {
-            throw BadRequestException.noTarget("Attribute " +
-                field + " does not have a value matching " +
-                "the filter " + valueFilter);
-          }
-          return;
+          continue;
+        }
+        nodeUpdated = true;
+
+
+        if (attributeValue.isObject() && value.isObject())
+        {
+          updateNode((ObjectNode) attributeValue, null, value);
+        }
+        else if (isDelete && attributeValue.isObject())
+        {
+          // Mark the current index as a child that must be deleted. We cannot
+          // delete this yet because it would modify the 'i' index, which we are
+          // currently iterating over.
+          indexesToDelete.add(i);
+        }
+        else
+        {
+          array.set(i, value);
         }
       }
-      updateNode(parent, field, value);
+
+      for (int j = indexesToDelete.size() - 1; j >= 0; j--)
```

Comment:
I think this looks good and doesn't necessarily need to change, but it seems possible to avoid the second loop for 'indexesToDelete' if we just go backwards through the array in the first loop. I think keeping the 'update' and 'delete' steps separated out might just be a cleaner implementation, though.

Reply:
That's a good idea, I didn't think of that. I initially had the main loop delete the index and then decrement `i` to account for the shifting, but I agree that it seems cleaner to just separate the two steps.